File: stml_py/tests/test_lexer.py
"""Unit tests for the STML lexer (via tokenize)."""

import stml


class TestLexer:
    """Lexer unit tests via stml.tokenize()."""

    def _token_types(self, tokens):
        """Extract token type names as a list."""
        return [t["type"] for t in tokens]

    def _token_values(self, tokens):
        """Extract token values as a list."""
        return [t["value"] for t in tokens]

    # ---- Basic structure ----

    def test_simple_key_value(self):
        tokens, _ = stml.tokenize('"key": "value"\n')
        types = self._token_types(tokens)
        assert "KEY" in types
        assert "COLON" in types
        assert "SCALAR" in types

    def test_null_value(self):
        tokens, _ = stml.tokenize('"key": null\n')
        types = self._token_types(tokens)
        assert "NULL_" in types

    def test_tilde_null(self):
        tokens, _ = stml.tokenize('"key": ~\n')
        types = self._token_types(tokens)
        assert "NULL_" in types

    # ---- Indentation ----

    def test_indent_dedent(self):
        text = '"root":\n  "child": "value"\n'
        tokens, _ = stml.tokenize(text)
        types = self._token_types(tokens)
        assert "INDENT" in types
        assert "DEDENT" in types

    def test_deep_nesting(self):
        text = '"a":\n  "b":\n    "c": "d"\n'
        tokens, _ = stml.tokenize(text)
        types = self._token_types(tokens)
        assert types.count("INDENT") == 2
        assert types.count("DEDENT") == 2

    # ---- Sequences ----

    def test_dash_sequence(self):
        text = '"list":\n  - "item1"\n  - "item2"\n'
        tokens, _ = stml.tokenize(text)
        types = self._token_types(tokens)
        assert types.count("DASH") == 2

    # ---- Quoted keys ----

    def test_quoted_key_non_greedy(self):
        text = '"ke\"y": "value"\n'
        tokens, _ = stml.tokenize(text)
        # The key should be ke"y (unescaped)
        for t in tokens:
            if t["type"] == "KEY":
                assert t["value"] == 'ke"y'

    # ---- Inline lists ----

    def test_inline_list(self):
        text = '"nums": [1, 2, 3]\n'
        tokens, _ = stml.tokenize(text)
        types = self._token_types(tokens)
        assert "INLINE_LIST" in types

    # ---- Multiline strings ----

    def test_multiline_string(self):
        text = '"text": {\n  line1\n  line2\n}\n'
        tokens, _ = stml.tokenize(text)
        types = self._token_types(tokens)
        assert "MULTILINE_STRING" in types

    # ---- Comments ----

    def test_comment_line(self):
        text = '# this is a comment\n"key": "value"\n'
        tokens, _ = stml.tokenize(text)
        # Comment line is a newline, not a separate token type
        types = self._token_types(tokens)
        assert "KEY" in types

    # ---- Edge cases ----

    def test_empty_input(self):
        tokens, _ = stml.tokenize("")
        # Should at least have END token
        assert len(tokens) >= 1
        assert tokens[-1]["type"] == "END"

    def test_unclosed_quote_produces_warning(self):
        text = '"key": "unclosed string\n'
        tokens, warnings = stml.tokenize(text)
        # Should produce at least one warning for unclosed quote
        raw_strings = [t for t in tokens if t["type"] == "RAW_STRING"]
        assert len(raw_strings) >= 1 or len(warnings) >= 1

File: stml_java/src/test/java/com/stml/TestLexer.java
package com.stml;

import com.stml.ast.*;
import com.stml.lexer.STMLLexer;
import com.stml.lexer.TokenType;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class TestLexer {

    @Test
    void testEmptyInput() {
        var result = Stml.tokenize("");
        assertFalse(result.tokens().isEmpty());
        // Should produce at least END
        assertTrue(result.tokens().stream().anyMatch(t -> t.type() == TokenType.END));
    }

    @Test
    void testSimpleKeyValue() {
        var result = Stml.tokenize("key: value");
        var tokens = result.tokens();
        assertTrue(tokens.stream().anyMatch(t -> t.type() == TokenType.KEY));
        assertTrue(tokens.stream().anyMatch(t -> t.type() == TokenType.COLON));
        assertTrue(tokens.stream().anyMatch(t -> t.type() == TokenType.SCALAR));
    }

    @Test
    void testQuotedKey() {
        var result = Stml.tokenize("\"my key\": value");
        var tokens = result.tokens();
        assertTrue(tokens.stream().anyMatch(t -> t.type() == TokenType.KEY));
    }

    @Test
    void testNullValue() {
        var result = Stml.tokenize("key: null");
        var tokens = result.tokens();
        assertTrue(tokens.stream().anyMatch(t -> t.type() == TokenType.NULL_));
    }

    @Test
    void testTildeNullValue() {
        var result = Stml.tokenize("key: ~");
        var tokens = result.tokens();
        assertTrue(tokens.stream().anyMatch(t -> t.type() == TokenType.NULL_));
    }

    @Test
    void testBareKey() {
        var result = Stml.tokenize("just_a_key");
        var tokens = result.tokens();
        assertTrue(tokens.stream().anyMatch(t -> t.type() == TokenType.BARE_KEY));
    }

    @Test
    void testSequenceEntry() {
        var result = Stml.tokenize("- item1");
        var tokens = result.tokens();
        assertTrue(tokens.stream().anyMatch(t -> t.type() == TokenType.DASH));
        assertTrue(tokens.stream().anyMatch(t -> t.type() == TokenType.SCALAR));
    }

    @Test
    void testInlineList() {
        var result = Stml.tokenize("key: [a, b, c]");
        var tokens = result.tokens();
        assertTrue(tokens.stream().anyMatch(t -> t.type() == TokenType.INLINE_LIST));
    }

    @Test
    void testDocSeparator() {
        var result = Stml.tokenize("---\nkey: value\n---\nkey2: value2");
        var tokens = result.tokens();
        long docSepCount = tokens.stream().filter(t -> t.type() == TokenType.DOC_SEPARATOR).count();
        assertEquals(2, docSepCount);
    }

    @Test
    void testIndentDedent() {
        var result = Stml.tokenize("key:\n  subkey: value");
        var tokens = result.tokens();
        assertTrue(tokens.stream().anyMatch(t -> t.type() == TokenType.INDENT));
        assertTrue(tokens.stream().anyMatch(t -> t.type() == TokenType.DEDENT));
    }

    @Test
    void testMultilineString() {
        var result = Stml.tokenize("key: {\n  line1\n  line2\n}");
        var tokens = result.tokens();
        assertTrue(tokens.stream().anyMatch(t -> t.type() == TokenType.MULTILINE_STRING));
    }

    @Test
    void testIteratorNotNull() {
        var result = Stml.tokenize("key: value");
        assertNotNull(result.tokens());
        assertNotNull(result.warnings());
    }
}
